crates/red_knot_python_semantic/resources/mdtest/unary/instance.md

```diff
@@ -30,3 +32,36 @@ b = NoDunder()
 -b  # error: [unsupported-operator] "Unary operator `-` is unsupported for type `NoDunder`"
 ~b  # error: [unsupported-operator] "Unary operator `~` is unsupported for type `NoDunder`"
 ```
+
+## Not
+
+Not operator is inferred based on <https://docs.python.org/3/library/stdtypes.html#truth-value-testing>
+Unless the `__bool__` or `__len__` is returning `False` it's always true.
+
+```py
+class AlwaysTrue:
+    def __bool__(self) -> Literal[True]:
+        return True
+
+
+# error: [invalid-method] "Method __bool__ for type `AlwaysTrue` returns type `@Todo` rather than `bool`"
+# revealed: bool
+reveal_type(not AlwaysTrue())
+
+
+class AlwaysFalse:
+    def __bool__(self) -> Literal[False]:
+        return False
+
+
+# error: [invalid-method] "Method __bool__ for type `AlwaysFalse` returns type `@Todo` rather than `bool`"
+# revealed: bool
+reveal_type(not AlwaysFalse())
+
+
+class NoBoolMethod: ...
+
+
+# revealed: Literal[False]
```

Comment:
Because everything is considered to be True in conditions unless specified by the __bool__ or __len__ method this will be false. The previous behavior was to say this is `bool`.

Reply:
Because it is permitted for a subclass to add a `__bool__` method when the base class doesn't have one, and `Type::Instance` types include subclasses, we cannot safely rely on this behavior. (When we do find a `__bool__` method, we can rely on this because it would be an invalid override for a subclass to override `__bool__` with a return type that isn't a subtype of the base class' return type.)

So we need to infer `bool` here.